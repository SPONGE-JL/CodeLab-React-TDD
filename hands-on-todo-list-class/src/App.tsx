import React, { Component } from 'react';
import Styled from 'styled-components';
import type { IScriptSnapshot } from 'typescript';

import { Button, Input, TodoItem } from 'Components';

//- Sytled Component
const Container = Styled.div`
  min-height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Contents = Styled.div`
  width: 50%;
  min-width: 400px;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  border-radius: 8px;
  background-color: #F2F2F2;
  box-shadow: 5px 5px 10px rgba(48, 36, 36, 0.2);
`;

const TodoItemListContainer = Styled.div`
  width: 100%;
  min-width: 350px;
  height: 50%;
  min-height: 400px;
  overflow: scroll;
  margin-bottom: 20px;
  border: 1px solid #BDBDBD;
  border-radius: 8px;
  background-color: #FFFFFF;
`;

const InputContainer = Styled.div`
  width: 100%;
  display: flex;
`;

//- React Component
interface AppProps {}

interface AppState {
  readonly error: boolean;
  readonly todo: string;
  readonly todoList: string[];
}

export class App extends Component<AppProps, AppState> {
  //- Initiation States
  // https://ko.reactjs.org/docs/react-component.html#constructor
  constructor(props: AppProps) {
    super(props);

    this.state = {
      error: false,
      todo: '',
      todoList: [],
    };
  }

  //- Render
  render() {
    const { todo, todoList, error } = this.state;
    return (
      <Container>
        {!error && (
          <Contents>
            <TodoItemListContainer data-testid="todoList">
              {todoList.map((toDoLabel, index) => (
                <TodoItem key={index} label={toDoLabel} onDelete={() => this.deleteTodo(index)} />
              ))}
            </TodoItemListContainer>
            <InputContainer>
              <Input placeholder="Insert a new todo" value={todo} onChange={(inputText) => this.setState({ todo: inputText })} />
              <Button label="Add" onClick={this.addTodo} />
            </InputContainer>
          </Contents>
        )}
      </Container>
    );
  }

  //- Event Functions
  private addTodo = (): void => {
    const { todo, todoList } = this.state;

    if (todo) {
      this.setState({
        todo: '',
        todoList: [...todoList, todo],
      });
    }
  };

  private deleteTodo = (index: number): void => {
    const { todoList } = this.state;
    let list = [...todoList];
    list.splice(index, 1);
    this.setState({ todoList: list });
  };

  //- Life-cycle Functions
  //? When craete the component:
  //> contructor -> getDerivedFromProps -> render -> componentDidMount
  //? When modify the props of component:
  //> getDerivedFromProps -> shouldComponentUpdate -> render -> getSnapshotBeforeUpdate -> componentDidUpdate
  //? When modify the state of componet:
  //> shduldcomponentUpdate -> render -> getSnapshotBeforUpdate -> componentDidUpdate
  //? When occured error in rendering:
  //> componentDidCatch
  //? When removed the component:
  //> componentWithUnmount

  // https://ko.reactjs.org/docs/react-component.html#componentdidmount
  componentDidMount() {
    console.log('exec > componentDidMount');
  }

  // https://ko.reactjs.org/docs/react-component.html#getsnapshotbeforeupdate
  getSnapshotBeforeUpdate(PrevProps: AppProps, prevState: AppState) {
    console.log('exec > getSnapshotBeforeUpdate');
    return { testData: true }; // serve to componentDidUpdate as snapshot.
  }

  // https://ko.reactjs.org/docs/react-component.html#componentdidupdate
  componentDidUpdate(prevProps: AppProps, prevState: AppState, snapshot: IScriptSnapshot) {
    console.log('exec > componentDidUpdate');
  }

  // https://ko.reactjs.org/docs/react-component.html#shouldcomponentupdate
  shouldComponentUpdate(nextProps: AppProps, nextState: AppState): boolean {
    console.log('exec > shouldComponentUpdate');
    return true; // true: re-render / false: block to re-render
    // return nextProps.id !== this.props.id;
  }

  // https://ko.reactjs.org/docs/react-component.html#unmounting
  componentWillUnmount() {
    console.log('exec > componentWillUnmount');
  }

  // https://ko.reactjs.org/docs/react-component.html#error-handling
  // Check > https://ko.reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html
  static getDerivedFromProps(nextProps: AppProps, prevState: AppState) {
    console.log('exec > getDerivedFromProps');
    return null;
    // return (nextProps.id !== prevState.id) ? { value: nextProps.value } : null
  }

  // https://ko.reactjs.org/docs/react-component.html#error-handling
  componentDidCatch(error: Error, info: React.ErrorInfo) {
    this.setState({ error: true });
  }
}

export default App;
