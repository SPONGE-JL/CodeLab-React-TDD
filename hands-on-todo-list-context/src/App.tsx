import React, { useState } from 'react';
import Styled from 'styled-components';

import { TodoList, InputBox } from 'Components';

//- Sytled Component
const Container = Styled.div`
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  text-align: center;
  justify-content: center;
`;

const Contents = Styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 50%;
  margin: 0% 25%;
  padding: 20px;
  border-radius: 8px;
  background-color: : #FFFFFF;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
`;

//- React Component
function App() {
  //- States
  const [todo, setTodo] = useState('');
  const [todoList, setTodoList] = useState<string[]>([]);

  //- Event Functions
  const addTodo = (): void => {
    if (todo) {
      setTodoList([...todoList, todo]);
      setTodo('');
    }
  };
  const deleteTodo = (index: number): void => {
    let list = [...todoList];
    list.splice(index, 1);
    setTodoList(list);
  };

  //- Render
  return (
    <Container>
      <Contents>
        <TodoList todoList={todoList} deleteTodo={deleteTodo} />
        <InputBox todo={todo} onType={(inputText) => setTodo(inputText)} onAdd={addTodo} />
      </Contents>
    </Container>
  );
}

export default App;
