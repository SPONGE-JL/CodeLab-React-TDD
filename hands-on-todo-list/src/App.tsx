import React, { useState } from 'react';
import Styled from 'styled-components';

import { Button, Input, ToDoItem } from 'Components';

//- Sytled Component
const Container = Styled.div`
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  text-align: center;
  justify-content: center;
`;

const Contents = Styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  border-radius: 8px;
  background-color: : #FFFFFF;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
`;

const ToDoItemListContainer = Styled.div`
  min-width: 350px;
  height: 400px;
  overflow: scroll;
  border: 1px solid #BDBDBD;
  border-radius: 8px;
  margin-bottom: 20px;
`;

const InputContainer = Styled.div`
  display: flex;
`;

//- React Component
function App() {
  //- States
  const [toDo, setToDo] = useState('');
  const [toDoList, setToDoList] = useState<string[]>([]);

  //- Event Functions
  const addToDo = (): void => {
    if (toDo) {
      setToDoList([...toDoList, toDo]);
      setToDo('');
    }
  };
  const deleteToDo = (index: number): void => {
    let list = [...toDoList];
    list.splice(index, 1);
    setToDoList(list);
  };

  //- Render
  return (
    <Container>
      <Contents>
        <ToDoItemListContainer>
          {toDoList.map((toDoLabel, index) => (
            <ToDoItem key={index} label={toDoLabel} onDelete={() => deleteToDo(index)} />
          ))}
        </ToDoItemListContainer>
        <InputContainer>
          <Input placeholder="Insert a new task" value={toDo} onChange={(inputText) => setToDo(inputText)} />
          <Button label="Add" onClick={addToDo} />
        </InputContainer>
      </Contents>
    </Container>
  );
}

export default App;
