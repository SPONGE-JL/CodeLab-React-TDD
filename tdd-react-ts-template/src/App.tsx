import React from 'react';
import Styled, { keyframes } from 'styled-components';

import logo from './logo.svg';

const AppContainer = Styled.div`
  text-align: center;
`;

const AppHeader = Styled.header`
  background-color: #282c34 !important;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
`;

const AppLogoSpin = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`;

const AppLogo = Styled.img`
  height: 40vmin;
  pointer-events: none;

  @media (prefers-reduced-motion: no-preference) {
    animation: ${AppLogoSpin} infinite 20s linear;
  }
}
`;

const AppLink = Styled.a`
  color: #61dafb;
`;

function App() {
  return (
    <AppContainer>
      <AppHeader>
        <AppLogo src={logo} alt ="logo" />
        <p>
          Edit <code>src/App.tsx</code> and save to reload.
        </p>
        <AppLink
          className="App-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          Learn React
        </AppLink>
      </AppHeader>
    </AppContainer>
  );
}

export default App;
