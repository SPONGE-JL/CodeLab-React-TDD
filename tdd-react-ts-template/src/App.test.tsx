import React from 'react';

import { render, screen } from '@testing-library/react';
import App from './App';

describe('<App />', () => {
  it('renders component correctly', () => {
    const { container } = render(<App />);

    // Check Link
    const linkElement = screen.getByText(/learn react/i);
    expect(linkElement).toBeInTheDocument();

    // Check Image Tag (App-logo)
    const appLogo = screen.getByAltText("logo");
    expect(appLogo).toBeInTheDocument();
    expect(appLogo).toHaveAttribute('src', 'logo.svg');

    // Check P Tag
    const elmByP = container.getElementsByTagName('p');
    expect(elmByP).toHaveLength(1);
    expect(elmByP[0]).toHaveTextContent('Edit src/App.tsx and save to reload');

    // Snapshot
    expect(container).toMatchSnapshot();
  });
});
