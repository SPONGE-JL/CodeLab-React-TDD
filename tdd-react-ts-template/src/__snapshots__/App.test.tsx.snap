// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders component correctly 1`] = `
<div>
  <div
    class="sc-bdfBQB dWImOz"
  >
    <header
      class="sc-gsTEea gPkLbi"
    >
      <img
        alt="logo"
        class="sc-dlfnuX kgqYpz"
        src="logo.svg"
      />
      <p>
        Edit 
        <code>
          src/App.tsx
        </code>
         and save to reload.
      </p>
      <a
        class="sc-hKgJUU cHcczu App-link"
        href="https://reactjs.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn React
      </a>
    </header>
  </div>
</div>
`;
